pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -306,7 +322,26 @@ func (e *EndpointController) onServiceDelete(obj interface{}) {
 	}
 
 	e.serviceSelectorCache.Delete(key)
-	e.queue.Add(key)
+	e.queueServiceSync(key, false /* allowRateLimiting */)
+}
+
+// queueServicSync schedules run of e.syncService for given key.
+// It use rate limiter to reduce number of generated endpoints objects if allowRateLimiting is true.
+func (e *EndpointController) queueServiceSync(key string, allowRateLimiting bool) {
```

Comment:
Not sure if it is worth refactor this into a separate struct that handles this. 

```
type rateLimitingQueue struct {
	queue workqueue.RateLimitingInterface		queue workqueue.RateLimitingInterface
	endpointUpdatesBatchPeriod time.Duration		// syncRateLimiter is used to track per-service rate limiter.
	syncRateLimiter *workqueue.ItemBucketRateLimiter
	queuedNotStarted     sets.String
	queuedNotStartedLock sync.Mutex
}

EnqueueService(key, allowRateLimiting) 
DequeueService() (key, quit)
Done()
Forget()

```

Reply:
This is to prevent `queuedNotStarted` from leaking everywhere. 

Plus, this smarter queue can be unit tested more thoroughly.

Reply:
Good idea. We will be able to reuse this struct in endpointslice controller as well.

Done

Reply:
I moved this part to https://github.com/kubernetes/kubernetes/pull/88579.

I will appreciate quick review.